services/app-api/src/postgres/contractAndRates/withdrawContract.ts

```diff
@@ -0,0 +1,319 @@
+import type { PrismaTransactionType } from '../prismaTypes'
+import type { ContractType } from '../../domain-models'
+import { unlockContractInsideTransaction } from './unlockContract'
+import { findStatePrograms } from '../state'
+import { packageName } from '@mc-review/hpp'
+import { submitContractInsideTransaction } from './submitContract'
+import { findContractWithHistory } from './findContractWithHistory'
+import { NotFoundError } from '../postgresErrors'
+import type { ExtendedPrismaClient } from '../prismaClient'
+
+export type WithdrawContractArgsType = {
+    contract: ContractType
+    updatedByID: string
+    updatedReason: string
+}
+
+/**
+ * Withdraws a contract and its associated eligible rates within a database transaction.
+ *
+ * This function performs the following operations:
+ * 1. Identifies related rates that should be withdrawn with the contract
+ * 2. Unlocks the contract and child rates with a withdrawal reason
+ * 3. Resubmits the contract and child rates with a withdrawal reason
+ * 4. Creates a withdrawal action for the contract
+ * 5. Creates withdrawal actions for all eligible associated rates
+ *
+ * @async
+ * @param {WithdrawContractArgsType} args - Withdrawal arguments
+ * @param {ContractType} args.contract - The contract to withdraw
+ * @param {string} args.updatedReason - The reason for withdrawal
+ * @param {string} args.updatedByID - ID of the user performing the withdrawal
+ *
+ */
+const withdrawContractInsideTransaction = async (
+    tx: PrismaTransactionType,
+    args: WithdrawContractArgsType
+): Promise<ContractType | Error> => {
+    performance.mark('beginWithdrawContractInsideTransaction')
```

Comment:
Do we usually leave this in here? Genuine question

Reply:
We don't! I missed this from clean up. Good catch.